File: 백준/Gold/2493. 탑/탑.java

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.Stack;
import java.util.StringTokenizer;

public class Main {
    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        Stack<Integer> s = new Stack<>();
        int N = Integer.parseInt(br.readLine());
        int[] heights = new int[N];
        int[] answer = new int[N];

        StringTokenizer st = new StringTokenizer(br.readLine());
        for (int i = 0; i < N; i++) {
            heights[i] = Integer.parseInt(st.nextToken());
        }

        for (int idx = N - 1; idx >= 0; idx--) {
            while (!s.isEmpty() && heights[idx] > heights[s.peek()]) {
                answer[s.pop()] = idx + 1;
            }

            s.add(idx);
        }

        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < answer.length; i++) {
            sb.append(answer[i]).append(" ");
        }

        System.out.println(sb);
    }
}

File: 백준/Gold/2493. 탑/탑.py
import heapq


N = int(input())
tops = [int(s) for s in input().split()]
answer = [0] * N
stack = []

for i in range(N - 1, -1, -1):
    while stack and tops[stack[-1]] < tops[i]:
        answer[stack.pop()] = i + 1
    stack.append(i)

print(' '.join(map(str, answer)))